First parseable meta@name=color-scheme should win

sideshowbarker opened the pull request:
#### 7d7215513fb5d3a7ac39d3809200c0a7aba58bd7
<pre>
First parseable meta@name=color-scheme should win
<a href="https://bugs.webkit.org/show_bug.cgi?id=213563">https://bugs.webkit.org/show_bug.cgi?id=213563</a>

Reviewed by Aditya Keerthi.

This change gets WebKit conforming to the requirement in the HTML spec
at <a href="https://html.spec.whatwg.org/#standard-metadata-names">https://html.spec.whatwg.org/#standard-metadata-names</a>:page&apos;s-supported-color-schemes-2
that UAs must respect the first parseable &lt;meta name=color-scheme&gt; value
encountered in tree order. And the change gets WebKit passing the WPT
tests related to that spec requirement.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-first-valid-applies-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-remove-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-remove-head-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::metaElementColorSchemeChanged):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMetaElement.cpp:
(WebCore::HTMLMetaElement::removedFromAncestor):
(WebCore::HTMLMetaElement::process):

Canonical link: <a href="https://commits.webkit.org/268064@main">https://commits.webkit.org/268064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de54749378ce906ef24c53d42719066b9b313386

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18590 "Failed to checkout and rebase branch from PR 17817") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19530 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20450 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17395 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18786 "Failed to checkout and rebase branch from PR 17817") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22238 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19070 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18815 "Failed to checkout and rebase branch from PR 17817") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/18975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/16182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/21326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/16944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/21326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/17218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/17114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/21326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/17711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/16765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/21133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2282 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17550 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4163,6 +4163,20 @@ void Document::updateViewportArguments()
 
+void Document::metaElementColorSchemeChanged()
+{
+    RefPtr<CSSValue> protector = nullptr;
```
`= nullptr` is unnecessary.

Though I think this would be better if the variable outside the loop was a `String` initialized to `emptyString()`, and then prior to breaking the loop, you could set it to the parsed value's `cssText`.

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4163,6 +4163,20 @@ void Document::updateViewportArguments()
+    if (protector)
+        processColorScheme(protector->cssText());
+    else
+        processColorScheme(emptyString());
```
Following the suggestion above, you would only need a single line with `processColorScheme`.

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4163,6 +4163,20 @@ void Document::updateViewportArguments()
+    RefPtr<CSSValue> protector = nullptr;
+    for (auto& metaElement : descendantsOfType<HTMLMetaElement>(rootNode())) {
+        const AtomString& nameValue = metaElement.attributeWithoutSynchronization(nameAttr);
+        if ((equalLettersIgnoringASCIICase(nameValue, "color-scheme"_s) || equalLettersIgnoringASCIICase(nameValue, "supported-color-schemes"_s)) && (protector = CSSParser::parseSingleValue(CSSPropertyColorScheme, metaElement.attributeWithoutSynchronization(contentAttr), CSSParserContext(document()))))
```
I think we should create the `CSSParserContext` once, rather than every loop iteration.

pxlcoder commented on Source/WebCore/dom/Document.h:
```diff
@@ -998,6 +998,8 @@ class Document
     void updateViewportArguments();
     void processReferrerPolicy(const String& policy, ReferrerPolicySource);
 
+    void metaElementColorSchemeChanged();
```
I think this should be behind `#if ENABLE(DARK_MODE_CSS)` like `processColorScheme`?

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4163,6 +4164,21 @@ void Document::updateViewportArguments()
+{
+    RefPtr<CSSValue> colorScheme;
+    String colorSchemeString = emptyString();
+    CSSParserContext cssParserContext = CSSParserContext(document());
```
```suggestion
    auto colorSchemeString = emptyString();
    auto cssParserContext = CSSParserContext(document());
```

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4163,6 +4164,21 @@ void Document::updateViewportArguments()
     }
 }
 
+void Document::metaElementColorSchemeChanged()
```
Will need to move this into `#if ENABLE(DARK_MODE_CSS)` too. There's a section slightly lower in this file.

pxlcoder commented on Source/WebCore/html/HTMLMetaElement.cpp:
```diff
@@ -143,6 +143,8 @@ void HTMLMetaElement::removedFromAncestor(RemovalType removalType, ContainerNode
 
     if (removalType.disconnectedFromDocument && equalLettersIgnoringASCIICase(name(), "theme-color"_s))
         oldParentOfRemovedTree.document().metaElementThemeColorChanged(*this);
+    else if (removalType.disconnectedFromDocument && (equalLettersIgnoringASCIICase(name(), "color-scheme"_s) || equalLettersIgnoringASCIICase(name(), "supported-color-schemes"_s)))
+        oldParentOfRemovedTree.document().metaElementColorSchemeChanged();
```
Will need to be in `#if ENABLE(DARK_MODE_CSS)` since this calls `metaElementColorSchemeChanged`.

pxlcoder reviewed (commented):
This is looking good, just a minor style comment + build fixes.

pxlcoder reviewed (approved):
Test failure is unrelated.

sideshowbarker committed:
First parseable meta@name=color-scheme should win
https://bugs.webkit.org/show_bug.cgi?id=213563

Reviewed by Aditya Keerthi.

This change gets WebKit conforming to the requirement in the HTML spec
at https://html.spec.whatwg.org/#standard-metadata-names:page's-supported-color-schemes-2
that UAs must respect the first parseable <meta name=color-scheme> value
encountered in tree order. And the change gets WebKit passing the WPT
tests related to that spec requirement.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-first-valid-applies-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-remove-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-remove-head-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::metaElementColorSchemeChanged):
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLMetaElement.cpp:
(WebCore::HTMLMetaElement::removedFromAncestor):
(WebCore::HTMLMetaElement::process):

Canonical link: https://commits.webkit.org/268064@main
